ons_alpha/private_media/signal_handlers.py

```diff
@@ -0,0 +1,104 @@
+from collections import defaultdict
+
+from django.apps import apps
+from django.contrib.contenttypes.models import ContentType
+from django.db.models import Q, QuerySet
+from wagtail.images.models import AbstractImage
+from wagtail.models import PAGE_MODEL_CLASSES, Page, ReferenceIndex
+from wagtail.signals import page_published, page_unpublished
+
+from .models import PrivateMediaCollectionMember
+
+
+def get_private_media_models():
+    return [m for m in apps.get_models() if issubclass(m, PrivateMediaCollectionMember)]
+
+
+def publish_page_media_on_publish(instance, **kwargs):  # pylint: disable=unused-argument
+    """
+    Signal handler to be connected to the 'page_published' signal for
+    all page types. It is responsible for identifying any privacy-controlled
+    media used by the page, and ensuring that it is also made public.
+    """
+    ids_by_ctype = defaultdict(list)
+    for ct_id, obj_id in get_all_page_media_ids(instance):
+        ids_by_ctype[ct_id].append(obj_id)
+
+    for ct_id, id_list in ids_by_ctype.items():
+        model_class = ContentType.objects.get_for_id(ct_id).model_class()
+        id_list = [model_class._meta.pk_field.to_python(id) for id in id_list]
+        private_objects_qs = model_class.objects.filter(id__in=id_list, is_private=True).select_related(
+            "collection", "last_public_collection", "last_private_collection"
+        )
+        if issubclass(model_class, AbstractImage):
+            private_objects_qs = private_objects_qs.prefetch_related("renditions")
+        model_class.objects.bulk_make_public(list(private_objects_qs))
```

Comment:
Question: What if an image is used on 2 pages? Currently, the first publish would make the image public - is that intended?

Reply:
Thanks for sense-checking @RealOrangeOne! Yes, that is the intended behaviour. By publishing a page, you also publish all media uses on that page. It's okay for a draft page to be referencing a mix of private and public media, but problematic for a published page to be referencing private media.